### PR TITLE
[PyOV] Remote API support for Python

### PR DESCRIPTION
**Pull request by jiwaszki**
### Details:
 - Extended `Core` class to interact with Remote API via `compile_model`, `create_context` and `get_default_context`.
 - Added `RemoteContext` and `RemoteTensor` classes to Python API.
 - Added `ClContext` and `ClImage2DTensor` classes placeholders.
 - Added `VADisplayWrapper`, `VAContext` and `VASurfaceTensor` classes.
 - Added utility function to raise `NotImplemented` errors.
 
**TODO list (with cc's):**
- [x] Decide which `VADisplayWrapper` constructors stay. `VADisplayWrapper(VADisplay device)` (which translates to `void*`) is quite necessary to pass memory pointers from/to external libraries. `VADisplayWrapper(std::string& device)` is not necessary as it introduces additional build dependency on libva itself. However, there are no standard libva bindings that expose such functionalities.
- [ ] Proposal: change of `create_tensor` to `create_device_tensor -> RemoteTensor` to make difference clear while API expose `create_host_tensor -> Tensor`. (cc @ilya-lavrenov @slyalin)
- [x] Question: move CL and libva related classes to specific submodules `openvino.intel_gpu` or keep it as-is in flat namespace? (cc @ilya-lavrenov @slyalin @akuporos)
- [x] Decide on how memory management should work. My draft proposal: expose in future PR as "read-only" via custom bindings to `cl_mem` in case of accessing `ClImage2DTensor`.
- [x] Add test cases.

### Tickets:
 - *129468 and 129469*

### CODE REVIEW

**Review by rkazants — COMMENTED** [review]
do we have tests to demonstrate work with remote API and GPU va surfaces?

**Comment by jiwaszki** [comment]
> do we have tests to demonstrate work with remote API and GPU va surfaces?

@rkazants currently no. Hopefully in a future there will be some kind of library to test against/use as a reference and provider of proper VADisplays. For now I have removed the dependency on constructors and leave only the approach that wraps around already existing pointer from libva.

File: src/bindings/python/src/pyopenvino/core/remote_context.hpp
Comment (ilya-lavrenov) [inline src/bindings/python/src/pyopenvino/core/remote_context.hpp:20]: we can use:
https://github.com/openvinotoolkit/openvino/blob/851bdb404c65396f0efb0d5e590ee2f279df4db5/src/plugins/intel_gpu/include/va/va.h#L5-L7
otherwise, who installs `va/va.h` ? Build machine can miss this package

File: src/bindings/python/src/pyopenvino/core/remote_context.hpp
Comment (ilya-lavrenov) [inline src/bindings/python/src/pyopenvino/core/remote_context.hpp:28]: what is the purpose of these wrappers?

File: src/bindings/python/src/pyopenvino/core/remote_context.hpp
Comment (ilya-lavrenov) [inline src/bindings/python/src/pyopenvino/core/remote_context.hpp:45]: looks like the code above (`ClContextWrapper`) does not depend on GPU. Why do we need condition for `PY_ENABLE_GPU` then?

File: src/bindings/python/tests/test_runtime/test_remote_api.py
Comment (ilya-lavrenov) [inline src/bindings/python/tests/test_runtime/test_remote_api.py:108]: I believe OpenVINO should not be `GPU` / `OpenCL` dependent.
If GPU plugin is not available, we can throw exception in runtime during access to GPU, while import of classes can work

File: tools/benchmark_tool/openvino/__init__.py
Comment (ilya-lavrenov) [inline tools/benchmark_tool/openvino/__init__.py:65]: why do we need to expose wrappers to users?

File: src/bindings/python/tests/test_runtime/test_remote_api.py
Comment (ilya-lavrenov) [inline src/bindings/python/tests/test_runtime/test_remote_api.py:1]: is it possible to see integration / interaction with other OpenCL libraries like `pyopencl`?
E.g. how can I create external OpenCL context, queue and submit work to GPU device using it?

File: src/bindings/python/tests/test_runtime/test_remote_api.py
Comment (ilya-lavrenov) [inline src/bindings/python/tests/test_runtime/test_remote_api.py:104]: could you please demonstrate how can I use `ClContext`, `ClImage2DTensor` and all other classes in proposed API? `from openvino import ClContext` does not demonstrate the usage.

**Comment by jiwaszki** [comment]
Some of checks are failing as older OMZ version is still on branch.

Required commit for this PR: https://github.com/openvinotoolkit/open_model_zoo/commit/ef62e728c04ba55b3c73fcd50cf1a41a55d4edf7

**Comment by Wovchena** [comment]
Just merge the needed change into your beanch

File: src/bindings/python/src/pyopenvino/core/remote_context.hpp
Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/core/remote_context.hpp]: Use of raw `void *` instead.

File: src/bindings/python/src/pyopenvino/core/remote_context.hpp
Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/core/remote_context.hpp:22]: To provide similar inheritance scheme to one that GPU plugin is having, without actually including derived classes as they strongly rely on CL/VA headers.

File: src/bindings/python/src/pyopenvino/core/remote_context.hpp
Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/core/remote_context.hpp:38]: Removed this condition all together.

File: src/bindings/python/tests/test_runtime/test_remote_api.py
Comment (jiwaszki) [reply src/bindings/python/tests/test_runtime/test_remote_api.py]: Done.

File: src/bindings/python/tests/test_runtime/test_remote_api.py
Comment (jiwaszki) [reply src/bindings/python/tests/test_runtime/test_remote_api.py:1]: This is the topic of further improvements. For now all OpenCL "layers" of Remote API are there as placeholders for future.

File: src/bindings/python/tests/test_runtime/test_remote_api.py
Comment (jiwaszki) [reply src/bindings/python/tests/test_runtime/test_remote_api.py]: This is the topic of further improvements. For now all OpenCL "layers" of Remote API are there as placeholders for future.

File: tools/benchmark_tool/openvino/__init__.py
Comment (jiwaszki) [reply tools/benchmark_tool/openvino/__init__.py]: Users should be aware of the pointer they are providing.

For OV it means reducing the potential errors as only `VADisplayWrapper` shall be accepted into the function. Not everything that translates to `void *`. Plus it gives space for future improvements without compromising API.

File: src/bindings/python/src/pyopenvino/core/core.cpp
Comment (vladimir-paramuzov) [inline src/bindings/python/src/pyopenvino/core/core.cpp:259]: There's no such method in c++ core API. Why do we need it in python? I'd expect that va context should be created using a wrapper which would convert some specific parameters to AnyMap, so the `create_context` method above can be used.

Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/core/core.cpp:259]: This is exactly what this function is doing. Automatically doing it for the users. If someone would use directly properties it is still okay -- any time the user can provide `create_context` with custom properties.

Would you expect this?
```python
_ = core.create_context("GPU", {key: value, ...})
# and
_ = core.create_context("GPU", display)
# or
_ = core.create_context("GPU", display, -1)
```
Here is my opinion: if I were to see it first time, this is unclear what context will be produced and what is the difference. Especially that one time dict is accepted, other times some class (and integer). It is not a case like `ov.Model` creation where different options always result in the same output.

From Python perspective the function would look like this:
```python
# Keyword only arguments will apply here, as properties are missing for hardcoded VA creation:
def create_context(self, device_name: str, *, properties:dict = {}, display: VADisplayWrapper, target_tile_id: int = -1):
    # and some logic to dispatch it correctly:
    if properties is not None:
        if display is not None or target_tile_id is not None:
            # handle with warning or error
        ...
    if display is not None and ...
    # and so on...
```
or to be dispatched version which does the same thing as two functions overloads (giving it the same name). However both functions will produce different output -- depending on arguments and properties.

For this function it is more clear to have `va` in the name as it clearly states the purpose. This is additional layer that covers hardcoded values. But for clarity it makes sense to create "helper" function with different name.

**Review by ilya-lavrenov — CHANGES_REQUESTED** [review]
Don't merge without explicit review from @ilya-lavrenov  @vladimir-paramuzov

File: src/bindings/python/src/pyopenvino/core/remote_context.cpp
Comment (ilya-lavrenov) [inline src/bindings/python/src/pyopenvino/core/remote_context.cpp:28]: Suggest to keep alignment with C / C++ API

File: src/bindings/python/src/pyopenvino/core/remote_context.hpp
Comment (ilya-lavrenov) [inline src/bindings/python/src/pyopenvino/core/remote_context.hpp:44]: do we need this wrapper?

File: src/bindings/python/tests/test_runtime/test_remote_api.py
Comment (ilya-lavrenov) [reply src/bindings/python/tests/test_runtime/test_remote_api.py:1]: do we need to expose `ClContext` and `ClImage2DTensor` in current release?
Looks like they don't provide any meaningful API now

File: src/bindings/python/src/pyopenvino/core/remote_context.hpp
Comment (ilya-lavrenov) [reply src/bindings/python/src/pyopenvino/core/remote_context.hpp:44]: if we cannot pass pointer, can we pass uint64 ?

File: src/bindings/python/src/pyopenvino/core/remote_tensor.cpp
Comment (ilya-lavrenov) [inline src/bindings/python/src/pyopenvino/core/remote_tensor.cpp:24]: can we check this method is tests?

File: src/bindings/python/src/pyopenvino/core/remote_context.cpp
Comment (ilya-lavrenov) [inline src/bindings/python/src/pyopenvino/core/remote_context.cpp:24]: can we check this method in tests?

File: src/bindings/python/src/pyopenvino/core/remote_context.cpp
Comment (ilya-lavrenov) [inline src/bindings/python/src/pyopenvino/core/remote_context.cpp:41]: can we test this method in tests ?

**Review by vladimir-paramuzov — APPROVED** [review]

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
Let's port to 2024.0 branch once this PR is merged

**Pull request merged**
